### DISCUSSION TITLE
Is it possible to run Storybook with multiple versions of React?

### DISCUSSION THREAD

**Discussion by pasho** [storybookjs/storybook#28645] [General]
Doing a big re-architecture of a monolith frontend. We are too big to safely upgrade React and other fundamental libraries across the whole app. It causes too many issues. Better to upgrade page by page. That means several versions of React need to co-exist in one app. 

I managed to achieve this by breaking down the monolith onto multiple npm packages. There is a shell package, which references individual page packages. Shell also has the central storybook instance, which imports all the stories from the other packages.

- node_modules
  - react 17
- packages
  - shell
    - storybook
  - pageA _@ react 17_
  - pageB _@ react 18_
    - node_modules
      - react 18

This all works correctly in webpack. It loads the right version depending on where the code lives. But storybook (`@storybook/react`, `webpack5`) seem to hard-wire react to always be from the root node_modules. I can see it adds webpack resolve aliases for react and react-dom. This means that anything expecting react 18 gets react 17.

So far I haven't found any way to resolve this. Has anyone done anything similar? Is it possible at all?

**Comment by p-desaintchamas** [comment]
No updates on this ?

**Comment by zachad** [comment]
Curious about this as well, and if anybody has any experience testing a component library against multiple versions of React within a single storybook.

**Comment by shilman** [comment]
Probably not the answer you're looking for, but I'd recommend checking out composition if you're not familiar with it already:

https://storybook.js.org/docs/sharing/storybook-composition

**Reply by zachad** [reply]
Thanks for your answer - that's not far off actually!
